### PR TITLE
add router

### PR DESCRIPTION
**Pull request by aliiohs**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: config/config_loader.go [hunk 1]
```diff
@@ -33,26 +33,35 @@ import (
...
 		log.Printf("[consumerInit] %#v", errCon)
 		consumerConfig = nil
 	}
+
 	if errPro := ProviderInit(confProFile); errPro != nil {
 		log.Printf("[providerInit] %#v", errPro)
 		providerConfig = nil
 	}
+
+	if errPro := RouterInit(confRouterFile); errPro != nil {
```
Comment (flycash) [inline config/config_loader.go]: You had initialized the routerConfig here, why do you try to initialize the routerConfig in initRouterUrl() method? I think initializing routerConfig here is enough? What do you think?

File: registry/directory/directory.go [hunk 2]
```diff
@@ -208,8 +234,19 @@ func (dir *registryDirectory) cacheInvoker(url common.URL) {
 
 //select the protocol invokers from the directory
 func (dir *registryDirectory) List(invocation protocol.Invocation) []protocol.Invoker {
-	//TODO:router
-	return dir.cacheInvokers
+	invokers := dir.cacheInvokers
+	localRouters := dir.Routers()
+
+	fmt.Println("========", len(localRouters))
```
Comment (flycash) [inline registry/directory/directory.go]: remove it

File: registry/directory/directory.go [hunk 3]
```diff
@@ -208,8 +234,19 @@ func (dir *registryDirectory) cacheInvoker(url common.URL) {
 
 //select the protocol invokers from the directory
 func (dir *registryDirectory) List(invocation protocol.Invocation) []protocol.Invoker {
```
Comment (flycash) [inline registry/directory/directory.go]: This method is similar to static_directory.List(invocation). Could you refactor them?

**Comment by flycash** [comment]
Please add more tests.

File: cluster/router/condition_router.go [hunk 4]
```diff
@@ -104,6 +104,14 @@ func newConditionRouter(url *common.URL) (*ConditionRouter, error) {
 	}, nil
 }
 
+func (c ConditionRouter) Priority() int64 {
+	return c.priority
+}
+
+func (c ConditionRouter) Url() common.URL {
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/router/condition_router.go]: you need a government answer as below

https://golang.org/doc/faq#methods_on_values_or_pointers

File: cluster/directory/base_directory.go [hunk 5]
```diff
@@ -18,30 +18,77 @@
...
 import (
-	"sync"
-)
-import (
+	"github.com/apache/dubbo-go/cluster"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/extension"
+	"github.com/dubbogo/gost/container"
 	"go.uber.org/atomic"
+	"sync"
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/directory/base_directory.go]: roger that

File: cluster/directory/static_directory.go [hunk 6]
```diff
@@ -18,8 +18,11 @@
 package directory
 
 import (
+	"fmt"
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/directory/static_directory.go]: roger that

File: cluster/directory/static_directory.go [hunk 7]
```diff
@@ -53,8 +56,18 @@ func (dir *staticDirectory) IsAvailable() bool {
 }
 
 func (dir *staticDirectory) List(invocation protocol.Invocation) []protocol.Invoker {
-	//TODO:Here should add router
-	return dir.invokers
+	invokers := dir.invokers
+	localRouters := dir.routers
+	fmt.Println("========", len(localRouters))
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/directory/static_directory.go]: will use logger.debug

File: cluster/directory/static_directory.go [hunk 8]
```diff
@@ -53,8 +56,18 @@ func (dir *staticDirectory) IsAvailable() bool {
...
+
+	if len(localRouters) > 0 {
+		for _, router := range localRouters {
+			if reflect.ValueOf(router.Url()).IsNil() || router.Url().GetParamBool(constant.RUNTIME_KEY, false) {
+				invokers = router.Route(invokers, *dir.ConsumerUrl, invocation)
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/directory/static_directory.go]: will break when find the router

File: cluster/router_chain.go [hunk 9]
```diff
@@ -0,0 +1,46 @@
+package cluster
+
+import (
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/protocol"
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/router_chain.go]: roger that

File: cluster/router_chain.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+	"sort"
+)
+
+type RouterChain struct {
+	routers        []Router
+	builtinRouters []Router
+	Invokers       []protocol.Invoker
+}
+
+func NewRouterChain(url common.URL) *RouterChain {
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/router_chain.go]: i am not sure,if useless will remove.

File: cluster/router_chain.go [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+	//return &RouterChain{
+	//	builtinRouters: routers,
+	//	routers:        routers,
+	//}
+	return nil
+}
+
+func (r RouterChain) AddRouters(routers []Router) {
+	r.routers = append(r.routers, routers...)
+	sort.Slice(r.routers, func(i, j int) bool {
```
[the remark this answers is not in the document]
Comment (zouyx) [reply cluster/router_chain.go]: agree,will change to SliceStable

File: common/url.go [hunk 12]
```diff
@@ -242,6 +242,36 @@ func (c URL) Key() string {
 	//return c.ServiceKey()
 }
 
+//todo
+func (c *URL) GetBackupUrls() []*URL {
```
[the remark this answers is not in the document]
Comment (zouyx) [reply common/url.go]: will add more UT

File: config/condition_router_config.go [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+)
+import (
+	"github.com/apache/dubbo-go/cluster/directory"
+	"github.com/apache/dubbo-go/common"
+	"github.com/apache/dubbo-go/common/constant"
+	"github.com/apache/dubbo-go/common/logger"
+)
+
+var (
+	mutex sync.Mutex
```
[the remark this answers is not in the document]
Comment (zouyx) [reply config/condition_router_config.go]: will change

File: config/condition_router_config.go [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+		return perrors.Errorf("ioutil.ReadFile(file:%s) = error:%v", confRouterFile, perrors.WithStack(err))
+	}
+	routerConfig = &ConditionRouterConfig{}
+
+	err = yaml.Unmarshal(confFileStream, routerConfig)
+	if err != nil {
+		return perrors.Errorf("yaml.Unmarshal() = error:%v", perrors.WithStack(err))
+	}
+
+	logger.Debugf("provider config{%#v}\n", routerConfig)
```
[the remark this answers is not in the document]
Comment (zouyx) [reply config/condition_router_config.go]: roger that

File: config/condition_router_config.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+	directory.RouterUrlSet.Add(initRouterUrl())
+	fmt.Println("=====", directory.RouterUrlSet.Size())
```
[the remark this answers is not in the document]
Comment (zouyx) [reply config/condition_router_config.go]: will use logger

File: config/condition_router_config.go [hunk 16]
```diff
@@ -0,0 +1,146 @@
...
+	return nil
+}
+
+func initRouterUrl() *common.URL {
+	mutex.Lock()
+	if routerConfig == nil {
+		confRouterFile := os.Getenv(constant.CONF_ROUTER_FILE_PATH)
+		err := RouterInit(confRouterFile)
```
[the remark this answers is not in the document]
Comment (zouyx) [reply config/condition_router_config.go]: agree

File: config/condition_router_config_test.go [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+ * limitations under the License.
+ */
+package config
+
+import (
+	"strings"
+	"testing"
+)
+
+func TestString(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (zouyx) [reply config/condition_router_config_test.go]: agree

File: config/config_loader.go (hunk 1 above)
Comment (zouyx) [reply config/config_loader.go]: agree

File: registry/directory/directory.go (hunk 2 above)
Comment (zouyx) [reply registry/directory/directory.go]: roger that

**Comment by zouyx** [comment]
@AlexStocks you can close this pr

**Comment by AlexStocks** [comment]
> @AlexStocks you can close this pr

okay.

**Pull request closed without merging**